src/extension/atlas/atlas_unwrap.cpp

```diff
@@ -0,0 +1,197 @@
+	uint64_t ic = rindices.size();
+
+	float eps = 1.19209290e-7F; // Taken from xatlas.h
+	for (int j = 0; j < ic / 3; j++) {
+		Vector3 p0 = rvertices[rindices[j * 3 + 0]];
+		Vector3 p1 = rvertices[rindices[j * 3 + 1]];
+		Vector3 p2 = rvertices[rindices[j * 3 + 2]];
+		
+		if ((p0 - p1).length_squared() < eps || (p1 - p2).length_squared() < eps || (p2 - p0).length_squared() < eps) {
+			continue;
+		}
+
+		indices.push_back(rindices[j * 3 + 0]);
+		indices.push_back(rindices[j * 3 + 1]);
+		indices.push_back(rindices[j * 3 + 2]);
+	}
+	
+	// set up input mesh
+	xatlas::MeshDecl input_mesh;
+	input_mesh.indexData = indices.ptr();
+	input_mesh.indexCount = indices.size();
+	input_mesh.indexFormat = xatlas::IndexFormat::UInt32;
+
+	input_mesh.vertexCount = vertices.size() / 3;
+	input_mesh.vertexPositionData = vertices.ptr();
+	input_mesh.vertexPositionStride = sizeof(float) * 3;
+	input_mesh.vertexNormalData = normals.ptr();
+	input_mesh.vertexNormalStride = sizeof(float) * 3;
+	input_mesh.vertexUvData = nullptr;
+	input_mesh.vertexUvStride = 0;
+
+	xatlas::ChartOptions chart_options;
+	chart_options.fixWinding = true;
+
+	ERR_FAIL_COND_V_MSG(texelSize <= 0.0f, false, "Texel size must be greater than 0.");
+
+	xatlas::PackOptions pack_options;
+	pack_options.padding = 1;
+	pack_options.maxChartSize = 1022; // Lightmap atlassing needs 2 for padding between meshes, so 4096-2
+	pack_options.blockAlign = true;
+	pack_options.texelsPerUnit = 1.0f / texelSize;
+
+	xatlas::Atlas *atlas = xatlas::Create();
+
+	xatlas::AddMeshError err = xatlas::AddMesh(atlas, input_mesh, 1);
+	ERR_FAIL_COND_V_MSG(err != xatlas::AddMeshError::Success, false, xatlas::StringForEnum(err));
+
+	xatlas::Generate(atlas, chart_options, pack_options);
+	
+	ERR_FAIL_COND_V_MSG(atlas->chartCount == 0, false, "No charts generated");
+	
+	float w = (float)(atlas->width);
+	float h = (float)(atlas->height);
+	
+	if (w == 0 || h == 0)
+	{
+		xatlas::Destroy(atlas);
+		ERR_FAIL_COND_V_MSG(w == 0 || h == 0, false, "could not bake because there is no area");
+	}
+
+	const xatlas::Mesh &output = atlas->meshes[0];
+	
+	mesh.clear_surfaces();
+	
+	Ref<SurfaceTool> surfaces_tools;
+	surfaces_tools.instantiate();
+	surfaces_tools->begin(Mesh::PRIMITIVE_TRIANGLES);
+	
+	vc = output.vertexCount;
+	ic = output.indexCount;
+	
+	PackedColorArray rcolors = arrays[Mesh::ARRAY_COLOR];
+	PackedVector2Array r_uv2 = arrays[Mesh::ARRAY_TEX_UV2];
+	PackedFloat32Array rtangent = arrays[Mesh::ARRAY_TANGENT];
+	PackedInt32Array rbones = arrays[Mesh::ARRAY_BONES];
+	PackedFloat32Array rweights = arrays[Mesh::ARRAY_WEIGHTS];
+	
+	for (int i = 0; i < ic; i += 3) {
+		
+		for (int j = 0; j < 3; j++) {
+			// Index of an original vertex corresponding to the generated one.
+			int vertex_id = output.vertexArray[output.indexArray[i + j]].xref;
+			
+			if (surface_format & Mesh::ARRAY_FORMAT_COLOR) {
+                surfaces_tools->set_color(rcolors[vertex_id]);
```

Comment:
I'd run clang format on this file even if this is meant to be mostly copy-pasted code but I assume you've made some changes here.

If you've made multiple changes that it is unrealistic to expect to copy-paste the engine source code over this again, that's fine to then keep the formatting mostly as-is. But if you have made a lot of changes already, I'd ask to remove the trailing comments on lines (our style is to put comments on their own lines before the line they comment on).